src/Symfony/Component/String/AbstractString.php

```diff
@@ -0,0 +1,555 @@
+        return $str;
+    }
+
+    /**
+     * @return static
+     */
+    abstract public function replace(string $from, string $to): self;
+
+    /**
+     * @param string|callable $to
+     *
+     * @return static
+     */
+    public function replaceMatches(string $from, $to): self
+    {
+        if ('g' === $pattern[-1]) {
+            $pattern = substr($pattern, 0, -1);
+        }
+
+        if ($this->ignoreCase) {
+            $pattern .= 'i';
+        }
+
+        if (\is_array($to) || $to instanceof \Closure) {
+            $to = static function (array $m) use ($to): string {
+                $to = $to($m);
+
+                if ('' !== $to && (!\is_string($to) || !preg_match('//u', $to))) {
+                    throw new InvalidArgumentException('Replace callback must return a valid UTF-8 string.');
+                }
+
+                return $to;
+            };
+        } elseif ('' !== $to && !preg_match('//u', $to)) {
+            throw new InvalidArgumentException('Invalid UTF-8 string.');
+        } else {
+            $replace = 'preg_replace';
+        }
+
+        set_error_handler(static function ($t, $m) { throw new InvalidArgumentException($m); });
+
+        try {
+            if (null === $string = $replace($from.'u', $to, $this->string)) {
+                $lastError = preg_last_error();
+
+                foreach (get_defined_constants(true)['pcre'] as $k => $v) {
+                    if ($lastError === $v && '_ERROR' === substr($k, -6)) {
+                        throw new RuntimeException('Matching failed with '.$k.'.');
+                    }
+                }
+
+                throw new RuntimeException('Matching failed with unknown error code.');
+            }
+        } finally {
+            restore_error_handler();
+        }
+
+        $str = clone $this;
+        $str->string = $string;
+
+        return $str;
+    }
+
+    /**
+     * @return static
+     */
+    abstract public function slice(int $start = 0, int $length = null): self;
+
+    /**
+     * @return static
+     */
+    abstract public function splice(string $replacement, int $start = 0, int $length = null): self;
+
+    /**
+     * @return static[]
+     */
+    abstract public function split(string $delimiter, int $limit = null): array;
+
+    abstract public function startsWith(string $prefix): bool;
+
+    /**
+     * @return static
+     */
+    public function title(bool $allWords = false): self
+    {
+        $str = clone $this;
+
+        if ($allWords) {
+            $str->string = preg_replace_callback('/\b./u', static function ($m) {
+                return mb_convert_case($m[0], MB_CASE_TITLE, 'UTF-8');
+            }, $str->string);
+        } else {
+            $firstChar = mb_substr($str->string, 0, 1, 'UTF-8');
+            $str->string = mb_convert_case($firstChar, MB_CASE_TITLE, 'UTF-8').substr($str->string, \strlen($firstChar));
+        }
+
+        return $str;
+    }
+
+    public function toBinary(): BinaryString
```

Comment:
Ah ok, that design choice (and things like `__toString` returning utf-8 by contract) does mean that this component will probably only ever work with Unicode and utf-8 though. That's a fair choice if made deliberately, but a limitation that maybe should be documented? Also, maybe this could be reflected in the component name (Utf8String?). That would also solve naming issues like whether classes need a `Utf8`-prefix (2 inline comments down from here) by setting a utf-8 'context' for the entire package.